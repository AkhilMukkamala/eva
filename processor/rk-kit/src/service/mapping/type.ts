import { IndexSignatureBase } from '../common';

export type ThemeMappingType = IndexSignatureBase & any;

export interface ComponentMappingMetaType {
  variants: IndexSignatureBase & any;
  states: string[];
}

export interface ComponentMappingType {
  meta: ComponentMappingMetaType;
  appearance: AppearanceType;
}

export interface AppearanceType {
  mapping?: MappingType;
  variant?: VariantGroupType;
}

export type VariantGroupType = IndexSignatureBase & any;

export interface MappingType {
  [key: string]: any;

  state?: StateType;
}

export type StateType = IndexSignatureBase & any;

