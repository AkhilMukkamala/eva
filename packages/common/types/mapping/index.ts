import { IndexSignatureBase } from '../common';

export type StyleMappingType = IndexSignatureBase & any;

export type ThemeMappingType = IndexSignatureBase & any;

export interface ComponentMappingMetaType {
  variants: IndexSignatureBase & any;
  states: string[];
}

export interface ComponentMappingType {
  meta: ComponentMappingMetaType;
  appearance: AppearanceType;
}

export interface AppearanceType {
  mapping?: MappingType;
  variant?: VariantGroupType;
}

export type VariantGroupType = IndexSignatureBase & any;

export interface MappingType {
  [key: string]: any;

  state?: StateType;
}

export type StateType = IndexSignatureBase & any;

export interface MappingMetaType {
  name: string;
  appearance: string;
  states: string[];
  variants: string[];
}
