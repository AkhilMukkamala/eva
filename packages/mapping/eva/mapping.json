{
  "Radio": {
    "meta": {
      "variants": {
        "status": [
          "error"
        ],
        "size": [
          "small",
          "big"
        ]
      },
      "states": [
        "checked",
        "disabled",
        "active"
      ]
    },
    "appearance": {
      "default": {
        "mapping": {
          "size": 36,
          "innerSize": 24,
          "highlightSize": 60,
          "borderWidth": 2,
          "borderColor": "gray-primary",
          "selectColor": "transparent",
          "highlightColor": "transparent",
          "state": {
            "active": {
              "borderColor": "gray-dark",
              "highlightColor": "gray-light"
            },
            "checked": {
              "borderColor": "blue-primary",
              "selectColor": "blue-primary"
            },
            "disabled": {
              "borderColor": "gray-light"
            },
            "checked.active": {
              "borderColor": "blue-dark"
            },
            "checked.disabled": {
              "selectColor": "gray-primary"
            }
          }
        },
        "variant": {
          "status": {
            "error": {
              "mapping": {
                "borderColor": "pink-primary",
                "state": {
                  "checked": {
                    "borderColor": "pink-primary",
                    "selectColor": "pink-primary"
                  },
                  "checked.active": {
                    "borderColor": "pink-primary"
                  }
                }
              }
            }
          },
          "size": {
            "big": {
              "mapping": {
                "size": 42,
                "innerSize": 28,
                "highlightSize": 70
              }
            },
            "small": {
              "mapping": {
                "size": 30,
                "innerSize": 20,
                "highlightSize": 50
              }
            }
          }
        }
      }
    }
  }
}
