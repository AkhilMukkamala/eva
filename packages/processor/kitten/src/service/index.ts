export * from './common';
export * from './mapping';
export * from './style';
export * from './test';

