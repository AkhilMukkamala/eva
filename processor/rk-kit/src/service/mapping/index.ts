export * from './type';
export * from './mapping.service';
