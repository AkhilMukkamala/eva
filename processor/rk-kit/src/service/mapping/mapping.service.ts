import { safe } from '../common';
import {
  ThemeMappingType,
  ComponentMappingType,
  AppearanceType,
  VariantGroupType,
  MappingType,
  StateType,
} from './type';

export const APPEARANCE_DEFAULT = 'default';

export function getComponentVariantGroups(mapping: ThemeMappingType,
                                          component: string): string[] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ComponentMappingType) => {
    return Object.keys(value.meta.variants);
  })
}

export function getComponentVariants(mapping: ThemeMappingType,
                                     component: string): string[][] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ComponentMappingType) => {
    const variantGroups = Object.keys(value.meta.variants);

    return variantGroups.map((group: string) => {
      return value.meta.variants[group];
    });
  })
}

export function getComponentStates(mapping: ThemeMappingType,
                                   component: string): string[] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ComponentMappingType) => {
    return value.meta.states;
  });
}

export function getStateAppearanceMapping(mapping: ThemeMappingType,
                                          component: string,
                                          appearance: string,
                                          state: string): any | undefined {

  const appearanceMapping = getAppearanceMapping(mapping, component, appearance);

  return safe(appearanceMapping, (value: MappingType) => {
    return getMappingState(value, state);
  });
}

export function getStatelessAppearanceMapping(mapping: ThemeMappingType,
                                              component: string,
                                              appearance: string): any | undefined {

  const appearanceMapping = getAppearanceMapping(mapping, component, appearance);

  return safe(appearanceMapping, (value: MappingType) => {
    const {state, ...params} = value;
    return params;
  });
}

export function getStateVariantMapping(mapping: ThemeMappingType,
                                       component: string,
                                       appearance: string,
                                       variant: string,
                                       state: string): any | undefined {

  const variantMapping = getVariantMapping(mapping, component, appearance, variant);

  return safe(variantMapping, (value: MappingType) => {
    return getMappingState(value, state);
  });
}

export function getStatelessVariantMapping(mapping: ThemeMappingType,
                                           component: string,
                                           appearance: string,
                                           variant: string): any | undefined {

  const variantMapping = getVariantMapping(mapping, component, appearance, variant);

  return safe(variantMapping, (value: MappingType) => {
    const {state, ...params} = value;
    return params;
  });
}

function getComponentMapping(mapping: ThemeMappingType,
                             component: string): ComponentMappingType | undefined {

  return mapping[component];
}

function getAppearance(mapping: ThemeMappingType,
                       component: string,
                       appearance: string): AppearanceType | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ComponentMappingType) => {
    return value.appearance[appearance];
  });
}

function getMappingState(mapping: MappingType,
                         state: string): StateType | undefined {

  return safe(mapping.state, (value: StateType) => {
    return value[state];
  });
}

function getAppearanceMapping(mapping: ThemeMappingType,
                              component: string,
                              appearance: string): MappingType | undefined {

  const appearanceConfig = getAppearance(mapping, component, appearance);

  return safe(appearanceConfig, (value: AppearanceType) => {
    return value.mapping;
  });
}

function getAppearanceVariantGroups(mapping: ThemeMappingType,
                                    component: string,
                                    appearance: string): VariantGroupType | undefined {

  const appearanceConfig = getAppearance(mapping, component, appearance);

  return safe(appearanceConfig, (value: AppearanceType) => {
    return value.variant;
  });
}

function getVariantMapping(mapping: ThemeMappingType,
                           component: string,
                           appearance: string,
                           variant: string): MappingType | undefined {

  const variantGroups = getAppearanceVariantGroups(mapping, component, appearance);

  const groupName = safe(variantGroups, value => {
    return Object.keys(value).find((group: string) => {
      return value[group][variant] !== undefined;
    });
  });

  return safe(groupName, value => {
    return safe(variantGroups[value][variant], value => value.mapping);
  });
}
