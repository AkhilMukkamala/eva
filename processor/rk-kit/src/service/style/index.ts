export * from './type';
export * from './style.service';
