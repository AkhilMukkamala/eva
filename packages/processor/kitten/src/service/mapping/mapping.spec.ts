import * as Service from './mapping.service';
import * as config from './mapping.spec.config';

describe('@mapping: service methods checks', () => {

  describe('* appearance', () => {

    describe('* default', () => {

      const appearance = 'default';

      it('* stateless', () => {
        const mapping = Service.getStatelessAppearanceMapping(config.mapping, 'Test', appearance);

        expect(mapping).toMatchSnapshot();
      });

      describe('* state', () => {

        it('* active', () => {
          const mapping = Service.getStateAppearanceMapping(config.mapping, 'Test', appearance, 'active');

          expect(mapping).toMatchSnapshot();
        });

        it('* undefined', () => {
          const mapping = Service.getStateAppearanceMapping(config.mapping, 'Test', appearance, 'undefined');

          expect(mapping).toBeUndefined();
        });

      });

      describe('* variant', () => {

        it('* success', () => {
          const mapping = Service.getStatelessVariantMapping(config.mapping, 'Test', appearance, 'success');

          expect(mapping).toMatchSnapshot();
        });

        it('* undefined', () => {
          const mapping = Service.getStatelessVariantMapping(config.mapping, 'Test', appearance, 'undefined');

          expect(mapping).toBeUndefined();
        });

        describe('* state', () => {

          it('* active', () => {
            const mapping = Service.getStateVariantMapping(config.mapping, 'Test', appearance, 'success', 'active');

            expect(mapping).toMatchSnapshot();
          });

          it('* undefined', () => {
            const mapping = Service.getStateVariantMapping(config.mapping, 'Test', appearance, 'success', 'undefined');

            expect(mapping).toBeUndefined();
          });

        });

      });

    });

    describe('* custom', () => {

      const appearance = 'custom';

      it('* stateless', () => {
        const mapping = Service.getStatelessAppearanceMapping(config.mapping, 'Test', appearance);

        expect(mapping).toMatchSnapshot();
      });

      describe('* state', () => {

        it('* active', () => {
          const mapping = Service.getStateAppearanceMapping(config.mapping, 'Test', appearance, 'active');

          expect(mapping).toMatchSnapshot();
        });

        it('* undefined', () => {
          const mapping = Service.getStateAppearanceMapping(config.mapping, 'Test', appearance, 'undefined');

          expect(mapping).toBeUndefined();
        });

      });

      describe('* variant', () => {

        it('* success', () => {
          const mapping = Service.getStatelessVariantMapping(config.mapping, 'Test', appearance, 'success');

          expect(mapping).toMatchSnapshot();
        });

        it('* undefined', () => {
          const mapping = Service.getStatelessVariantMapping(config.mapping, 'Test', appearance, 'undefined');

          expect(mapping).toMatchSnapshot();
        });

        describe('* state', () => {

          it('* active', () => {
            const mapping = Service.getStateVariantMapping(config.mapping, 'Test', appearance, 'success', 'active');

            expect(mapping).toMatchSnapshot();
          });

          it('* undefined', () => {
            const mapping = Service.getStateVariantMapping(config.mapping, 'Test', appearance, 'success', 'undefined');

            expect(mapping).toBeUndefined();
          });

        });

      });

    });

    describe('* undefined', () => {

      const appearance = 'undefined';

      it('* stateless', () => {
        const mapping = Service.getStatelessAppearanceMapping(config.mapping, 'Test', appearance);

        expect(mapping).toBeUndefined();
      });

      describe('* state', () => {

        it('* active', () => {
          const mapping = Service.getStateAppearanceMapping(config.mapping, 'Test', appearance, 'active');

          expect(mapping).toBeUndefined();
        });

        it('* undefined', () => {
          const mapping = Service.getStateAppearanceMapping(config.mapping, 'Test', appearance, 'undefined');

          expect(mapping).toBeUndefined();
        });

      });

      describe('* variant', () => {

        it('* success', () => {
          const mapping = Service.getStatelessVariantMapping(config.mapping, 'Test', appearance, 'success');

          expect(mapping).toBeUndefined();
        });

        it('* undefined', () => {
          const mapping = Service.getStatelessVariantMapping(config.mapping, 'Test', appearance, 'undefined');

          expect(mapping).toBeUndefined();
        });

        describe('* state', () => {

          it('* active', () => {
            const mapping = Service.getStateVariantMapping(config.mapping, 'Test', appearance, 'success', 'active');

            expect(mapping).toBeUndefined();
          });

          it('* undefined', () => {
            const mapping = Service.getStateVariantMapping(config.mapping, 'Test', appearance, 'success', 'undefined');

            expect(mapping).toBeUndefined();
          });

        });

      });

    });

  });

  describe('* variants groups getting', () => {

    it('* expected component', () => {
      const value: string[] = Service.getComponentVariantGroups(config.mapping, 'Test');
      expect(value).toEqual(['status', 'size']);
    });

    it('* unexpected component', () => {
      const value: string[] | undefined = Service.getComponentVariantGroups(config.mapping, 'Undefined');
      expect(value).toBeUndefined();
    });

  });

});
