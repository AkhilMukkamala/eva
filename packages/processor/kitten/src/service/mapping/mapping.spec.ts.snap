// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@mapping: service methods checks * appearance * custom * state * active 1`] = `
Object {
  "borderColor": "grayLight",
}
`;

exports[`@mapping: service methods checks * appearance * custom * stateless 1`] = `
Object {
  "borderWidth": 4,
}
`;

exports[`@mapping: service methods checks * appearance * custom * variant * state * active 1`] = `undefined`;

exports[`@mapping: service methods checks * appearance * custom * variant * success 1`] = `
Object {
  "borderColor": "tealPrimary",
}
`;

exports[`@mapping: service methods checks * appearance * custom * variant * undefined 1`] = `undefined`;

exports[`@mapping: service methods checks * appearance * default * state * active 1`] = `
Object {
  "borderColor": "grayDark",
}
`;

exports[`@mapping: service methods checks * appearance * default * stateless 1`] = `
Object {
  "borderColor": "grayPrimary",
  "borderWidth": 2,
  "innerSize": 24,
  "selectColor": "transparent",
  "size": 36,
}
`;

exports[`@mapping: service methods checks * appearance * default * variant * state * active 1`] = `undefined`;

exports[`@mapping: service methods checks * appearance * default * variant * success 1`] = `Object {}`;
